server/src/main/java/org/tctalent/server/service/db/impl/DuolingoCouponServiceImpl.java

```diff
@@ -0,0 +1,206 @@
+
+package org.tctalent.server.service.db.impl;
+
+import com.opencsv.CSVReader;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.*;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.web.multipart.MultipartFile;
+import org.tctalent.server.model.db.DuolingoCoupon;
+import org.tctalent.server.repository.db.CandidateRepository;
+import org.tctalent.server.repository.db.DuolingoCouponRepository;
+import org.tctalent.server.response.DuolingoCouponResponse;
+import org.tctalent.server.service.db.DuolingoCouponService;
+
+@Service
+public class DuolingoCouponServiceImpl implements DuolingoCouponService {
+
+  private final DuolingoCouponRepository couponRepository;
+  private final CandidateRepository candidateRepository;
+
+  private static final DateTimeFormatter FORMATTER1 = DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss");
+  private static final DateTimeFormatter FORMATTER2 = DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm");
+
+  @Autowired
+  public DuolingoCouponServiceImpl(DuolingoCouponRepository couponRepository, CandidateRepository candidateRepository) {
+    this.couponRepository = couponRepository;
+    this.candidateRepository = candidateRepository;
+  }
+
+  /**
+   * Imports coupons from a CSV file, avoiding duplicates.
+   *
+   * @param file the uploaded CSV file containing coupon data
+   */
+  @Override
+  @Transactional
+  public void importCoupons(MultipartFile file) {
+    List<DuolingoCoupon> newCoupons = new ArrayList<>();
+    Set<String> seenCouponCodes = new HashSet<>();
+
+    try (CSVReader reader = new CSVReader(new InputStreamReader(file.getInputStream(), StandardCharsets.UTF_8))) {
+      reader.readNext(); // Skip header row
+      String[] line;
+
+      while ((line = reader.readNext()) != null) {
+        if (line.length < 6) continue; // Skip rows with insufficient data
+
+        String couponCode = line[0];
+        if (seenCouponCodes.contains(couponCode)) continue; // Skip if already processed
+        seenCouponCodes.add(couponCode); // Mark code as processed
+
+        if (!couponRepository.existsByCouponCode(couponCode)) {
+          DuolingoCoupon coupon = new DuolingoCoupon();
+          coupon.setCouponCode(couponCode);
+          coupon.setAssigneeEmail(getNullableValue(line[1]));
+          coupon.setExpirationDate(parseDate(line[2], FORMATTER1, FORMATTER2));
+          coupon.setDateSent(parseDate(line[3], FORMATTER1, FORMATTER2));
+          coupon.setCouponStatus(getNullableValue(line[4]));
+          coupon.setTestStatus(getNullableValue(line[5]));
+          newCoupons.add(coupon);
+        }
+      }
+      if (!newCoupons.isEmpty()) couponRepository.saveAll(newCoupons);
```

Comment:
Can you use braces with the body on a separate line? There's nothing wrong with how you've expressed it, but below is the general way it's been done throughout the project:
 
```suggestion
      if (!newCoupons.isEmpty()) {
        couponRepository.saveAll(newCoupons);
      }
```